### PR TITLE
Refactor compiler error handling

### LINKED ISSUE #2734: Use `Handler` infra instead of `CompileResult`
The idea is to stop using the `CompileResult<T>` monad in `sway_core`. The type is a beautiful concept, and would be the go-to solution in Haskell. However, there is lacking proper language support for early returns (i.e. `do` notation in Haskell). Meanwhile, the `?` operator is inappropriate for logging (writer monad), which is half of what `CompileResult` is (roughly `(Maybe a, [Warning], [Error])`).

Currently, use of `CompileResult<T>` results in a lot of noise dealing with error handling. For example, there's a bunch of code dealing with extending errors and warnings.

Instead of `CompileResult`, we can use the `Handler` concept in the parser, which is also the way [rustc deals with emitting errors](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/struct.Handler.html).

### PR DESCRIPTION
**Pull request by IGI-111**
## Description

Change error handling to use `Handler` instead of the macro based `CompileResult`.

Fix #2734

## Checklist

- [x] I have linked to any relevant issues.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have updated the documentation where relevant (API docs, the reference, and the Sway book).
- [x] I have added tests that prove my fix is effective or that my feature works.
- [x] I have added (or requested a maintainer to add) the necessary `Breaking*` or `New Feature` labels where relevant.
- [x] I have done my best to ensure that my PR adheres to [the Fuel Labs Code Review Standards](https://github.com/FuelLabs/rfcs/blob/master/text/code-standards/external-contributors.md).
- [x] I have requested a review from the relevant team or maintainers.

### CODE REVIEW

**Comment by anton-trunov** [comment]
This is supposed to fix #2734, right?

**Review by anton-trunov — COMMENTED** [review]
Wow, this is a lot of work! Awesome stuff.

I left several comments in the code and let me ask this one here. There is a new idiom with a mutable `error_emitted` variable. Could we somehow move this functionality into the handler? For instance, the user code calls a `handler.reset_new_error_emitted()` method and then checks `handler.is_new_error_emitted()` at the end, or something like this. I understand it might be a bit too late for this :)

File: test/src/e2e_vm_tests/test_programs/should_fail/where_clause_impls/test.toml [hunk 1]
```diff
@@ -1,11 +1,10 @@
...
+
 # check: $()x: a.x.my_add(b.x),
 # nextln: $()No method named "my_add" found for type "T".
 # check: $()y: a.y.my_add(b.y),
 
 # check: $()y: a.y.my_add(b.y),
 # nextln: $()No method named "my_add" found for type "T".
-
-# check: $()let baz = foo.my_add(bar);
-# nextln: $()No method named "my_add" found for type "MyPoint<u32>".
```
[the remark this answers is not in the document]
Comment (IGI-111) [reply test/src/e2e_vm_tests/test_programs/should_fail/where_clause_impls/test.toml:11]: There are some slight changes to control flow because refactoring things makes some functions infallible, allows us to detect cases where errors always/never happen.

In this case we're bailing because of the previous errors, which I think is reasonable given how the test code is broken.

If you repair the impl in the test so that the previous two errors are fixed, this error will show up, which I think is decent enough.

File: sway-core/src/semantic_analysis/ast_node/expression/typed_expression/if_expression.rs [hunk 2]
```diff
@@ -29,19 +30,21 @@ pub(crate) fn instantiate_if_expression(
...
-                then.return_type,
-                ty_to_check,
-                ctx.self_type(),
-                &then.span,
-                "`then` branch must return expected type.",
-                None
-            ),
-            warnings,
-            errors
+        let (warnings, errors) = type_engine.unify_with_self(
```
[the remark this answers is not in the document]
Comment (IGI-111) [reply sway-core/src/semantic_analysis/ast_node/expression/typed_expression/if_expression.rs:33]: Yeah I had an earlier version that included those, but since it isn't absolutely necessary and the change is already huge, I decided against refactoring those.

Should probably follow up with that once this is merged.

File: sway-error/src/handler.rs [hunk 3]
```diff
@@ -21,26 +22,94 @@ struct HandlerInner {
...
+        inner.errors = dedup_unsorted(inner.errors.clone());
+        inner.warnings = dedup_unsorted(inner.warnings.clone());
+    }
 }
 
 /// Proof that an error was emitted through a `Handler`.
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct ErrorEmitted {
-    _priv: (),
+pub struct ErrorEmitted;
```
[the remark this answers is not in the document]
Comment (IGI-111) [reply sway-error/src/handler.rs:75]: I figured it might have been the intent, but:

1. It doesn't stop you from creating instances at all
2. We actually do need the ability to make up some error emission to bail in the case of `ErrorRecovery` types

I would definitely want to improve our type resolution error recovery mechanism not to have to do this, but I think that's out of scope for this.

File: test/src/e2e_vm_tests/test_programs/should_fail/method_requires_mut_var/test.toml [hunk 4]
```diff
@@ -1,5 +1,6 @@
 category = "fail"
 
+# check: $()self.a.f();
```
[the remark this answers is not in the document]
Comment (IGI-111) [reply test/src/e2e_vm_tests/test_programs/should_fail/method_requires_mut_var/test.toml:3]: Slight changes in control flow allow more (correct) warnings to be produced, and some include this line which gets picked up by the test.

File: test/src/e2e_vm_tests/test_programs/should_fail/type_alias_private/test.toml [hunk 5]
```diff
@@ -1,4 +1,8 @@
 category = "fail"
 
+# check: $()fn foo(x: lib::Alias) {}
+
+# check: $()fn foo(x: lib::Alias) {}
+
```
[the remark this answers is not in the document]
Comment (IGI-111) [reply test/src/e2e_vm_tests/test_programs/should_fail/type_alias_private/test.toml:6]: Same answer, more warnings get produced so we have to delay the check.

**Comment by IGI-111** [comment]
> I left several comments in the code and let me ask this one here. There is a new idiom with a mutable `error_emitted` variable. Could we somehow move this functionality into the handler? For instance, the user code calls a `handler.reset_new_error_emitted()` method and then checks `handler.is_new_error_emitted()` at the end, or something like this. I understand it might be a bit too late for this :)

Yeah I think that merits some factoring but I'm not really sure what would be a Rusty way of doing this. Perhaps some sort of collection guard would do the trick.

Creating a local handler, checking it for errors and appending it to the top level one would definitely work but it might be error prone. I think we can do it more clearly.

That said I think that's out of scope for this.
This and the unify refactoring should be their own PRs.

**Comment by andrewvious** [comment]
Does this also close #4818 ?

**Comment by ironcev** [comment]
> Does this also close #4818 ?

Actually it doesn't. This PR changes the way how errors are reported, and not how they are defined. All the redundant spans are still there in the _error.rs_ and provided during the creation of errors. But #4818 is anyhow not that relevant at the moment because it heavily depends on the outcome of #21. Once #21 is done (I am just a few commits away from the PR :-)) we will decide how to proceed with the refactoring of errors which could actually render #4818 obsolete.

I'll comment this on #4818 as well and add the dependency to #21.

File: sway-error/src/handler.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ironcev) [reply sway-error/src/handler.rs:75]: TL;DR; Let's extend the `Handler` to clearly support different use cases that we can encounter in error reporting, rather then letting every piece of code freely instantiate `ErrorEmitted` instances. 

Ad 1) I am a bit confused. How could have the instances be created outside of the `Handler`? We had a private field that could not be set outside of the module, means `ErrorEmitted` instance could have been created only inside of the module, effectively only in the `Handler`. Or am I missing something?

Ad 2) I've seen this `ErrorRecovery` use case. Yet, I wouldn't solve it by allowing uncontrolled creating of `ErrorEmitted` instances. Here is why:
- with the new `Handler` and the way it is passed down we are getting a clear pattern for error reporting. If we have particular use cases in the error reporting, `Handler` should explicitly support it and offer a pattern for it. E.g. in this case something like `handler.emit_silent_err()` or something. (The same reasoning goes for `error_emitted` variable but I'll comment on that in the PR.)
- if we allow creation of `ErrorEmitted`, sooner or later they will start creeping through the code and be used for various workarounds bypassing the `Handler` and causing who knows what kind of side effects. Guaranteed :-)

**Comment by ironcev** [comment]
Wow, this was a humongous change! Honestly, I gave up somewhere in the middle. But I made sure to take a look at the interesting stuff :smile: 

Most of the changes are straightforward, kicking out the `CompileResult`, and consistently passing the `Handler` down. It's great to see how simpler the code is now, how more expressive the error reporting is. No more need for manual work, creating the error and warning vectors internally and such. Great!

Exactly because of this fact, that we are getting a clean pattern for error reporting I see @anton-trunov's remarks as extremely important. It would be a pity to miss the opportunity to get a clean solution for these two additional use cases that have emerged during the refactoring.

`error_emitted` is an imperative juggling that definitely requires a clear support in the `Handler` no matter how we implement that support. `Handler` API should made it explicit and remove the possibility of unintentional misuse. E.g., something along the lines (I am sure we can think of better approaches but just to demonstrate the idea of gaining safety through the type system/language features/...):

```
... handler ... // Outer/Parent handler.
{
    let  handler = handler.scoped(); // Outer handler cannot be used any more.

   ... handler ... // Pass the scoped one down. 

   if (handler.has_errors_in_scope()) {
      ...
   }   
} // Scoped handler propagates the collect errors/warnings to the parent. 
```

Same thing is with the `TypeRecovery`, as discussed in the code comment. It should be an explicitly supported pattern.

```
=> Err(handler.i_need_a_good_name_here())
``` 

And one last trivia. `handler.has_error()` and `handler.has_warning()` confuses the mathematician in me. It sounds like _has exactly one error/warning_ :smile:
`handler.has_error**s**()` and `handler.has_warning**s**()` would be more appropriate IMO.

Long story short, my proposal;
- [ ] Open a new issue for refactoring `error_emitted`.
- [ ] Create explicit handling for `TypeRecovery` and similar cases as a part of this PR.
- [ ] Humor me and rename the `handler.has_error()` and `handler.has_warning()` in this PR :smile:

I can gladly take over the last two points.

File: test/src/e2e_vm_tests/test_programs/should_fail/where_clause_impls/test.toml (hunk 1 above)
[the remark this answers is not in the document]
Comment (anton-trunov) [reply test/src/e2e_vm_tests/test_programs/should_fail/where_clause_impls/test.toml:11]: I think this change is sufficiently small not to pay too much attention to it, given the size of this refactoring :)

**Review by anton-trunov — APPROVED** [review]
I totally agree that we should merge this PR as it as, given its size. Merging it asap will let us proceed with other PRs without having to deal with lots of merge conflicts.

**Review by ironcev — APPROVED** [review]
Agree. I can definitely live with the above points being open for a while. And also I want to open PR for #21 for which these changes should already be integrated (code conflicts, I still want to change some errors and warnings as additional examples, ...). Considering the impact on the overall code base let's integrate it ASAP.

**Comment by IGI-111** [comment]
Alright, I'll merge this as is and create some issues for those remaining points:

* refactoring `error_emitted`
* finding a better facility for type system error recovery

I'll refactor the unifier and handler method names in a direct follow up.

**Comment by IGI-111** [comment]
For reference:
* #4885 
* #4886
